runtime/compiler/x/codegen/J9TreeEvaluator.cpp

```diff
@@ -11924,6 +11930,159 @@ J9::X86::TreeEvaluator::directCallEvaluator(TR::Node *node, TR::CodeGenerator *c
+   generateRegRegInstruction(TR::InstOpCode::PUNPCKHBWRegReg, node, xmmLowReg, zeroReg, cg);
+   generateMemRegInstruction(TR::InstOpCode::MOVDQUMemReg, node, generateX86MemoryReference(destBufferReg, destOffsetReg, 0, headerOffsetConst + vectorLengthConst, cg), xmmLowReg, cg);
+
+   generateRegRegInstruction(TR::InstOpCode::PUNPCKLBWRegReg, node, xmmHighReg, zeroReg, cg);
+   generateMemRegInstruction(TR::InstOpCode::MOVDQUMemReg, node, generateX86MemoryReference(destBufferReg, destOffsetReg, 0, headerOffsetConst, cg), xmmHighReg, cg);
+
+   // increase src offset by size of imm register
+   // increase dest offset by double, to account for the byte->char inflation
+   generateRegImmInstruction(TR::InstOpCode::ADD4RegImm4, node, srcOffsetReg, vectorLengthConst, cg);
+   generateRegImmInstruction(TR::InstOpCode::ADD4RegImm4, node, destOffsetReg, 2 * vectorLengthConst, cg);
+
+   // LOOP BACK
+   generateLabelInstruction(TR::InstOpCode::JMP4, node, startLoop, cg);
+   generateLabelInstruction(TR::InstOpCode::label, node, endLoop, cg);
+
+   // AND length with 15 to compute residual remainder
+   // then copy and interleave 8 bytes from src buffer with 0s into dest buffer if possible
+   generateRegImmInstruction(TR::InstOpCode::AND4RegImm4, node, lengthReg, vectorLengthConst - 1, cg);
+
+   generateRegImmInstruction(TR::InstOpCode::CMP4RegImm4, node, lengthReg, 8, cg);
+   generateLabelInstruction(TR::InstOpCode::JL1, node, afterCopy8Label, cg);
+
+   generateRegMemInstruction(TR::InstOpCode::MOVQRegMem, node, xmmLowReg, generateX86MemoryReference(srcBufferReg, srcOffsetReg, 0, headerOffsetConst, cg), cg);
+   generateRegRegInstruction(TR::InstOpCode::PUNPCKLBWRegReg, node, xmmLowReg, zeroReg, cg);
+   generateMemRegInstruction(TR::InstOpCode::MOVDQUMemReg, node, generateX86MemoryReference(destBufferReg, destOffsetReg, 0, headerOffsetConst, cg), xmmLowReg, cg);
+   generateRegImmInstruction(TR::InstOpCode::SUB4RegImm4, node, lengthReg, 8, cg);
+
+   generateRegImmInstruction(TR::InstOpCode::ADD4RegImm4, node, srcOffsetReg, 8, cg);
+   generateRegImmInstruction(TR::InstOpCode::ADD4RegImm4, node, destOffsetReg, 16, cg);
+
+   generateLabelInstruction(TR::InstOpCode::label, node, afterCopy8Label, cg);
+
+   // handle residual (< 8 bytes left) & jump to copy instructions based on the number of bytes left
+   // calculate how many bytes to skip based on length;
+
+   const int copy_instruction_size = 5  // size of MOVZXReg2Mem1
+                                    +4; // size of S2MemReg
+
+   // since copy_instruction_size could change depending on which registers are allocated to scratchReg, srcOffsetReg and destOffsetReg
+   // we reserve them to be eax, ecx, edx, respectively
+
+   generateRegRegImmInstruction(TR::InstOpCode::IMUL4RegRegImm4, node, lengthReg, lengthReg, -copy_instruction_size, cg);
+   generateRegImmInstruction(TR::InstOpCode::ADD4RegImm4, node, lengthReg, copy_instruction_size * 7, cg);
```

Comment:
Not sure if there is a better way to do this. Seems fragile to changes but at least we have the comment explaining why it is so.

Reply:
Yeah, there really is no way to know the size of the instructions without reserving the registers because the REX prefix might be used. Maybe later we could modify the register dependencies to request any register that does not require the rex prefix.